django/db/models/aggregates.py

```diff
@@ -211,3 +213,39 @@ def __init__(self, expression, sample=False, **extra):
 
     def _get_repr_options(self):
         return {**super()._get_repr_options(), "sample": self.function == "VAR_SAMP"}
+
+
+class JSONArrayAgg(Aggregate):
+    function = "JSON_ARRAYAGG"
+    output_field = JSONField()
+    name = "JSONArrayAgg"
+    arity = 1
+
+    def as_sqlite(self, compiler, connection, **extra_context):
+        sql, params = self.as_sql(
+            compiler, connection, function="JSON_GROUP_ARRAY", **extra_context
+        )
+        if (default := self.default) == []:
+            return sql, params
+        count = self.copy()
+        count.__class__ = Count
+        count_sql, count_params = compiler.compile(count)
+        default_sql = ""
+        default_params = () if self.filter is not None else []
+        if default is not None:
+            default_sql, default_params = compiler.compile(default)
+            default_sql = f" ELSE {default_sql}"
+        sql = f"(CASE WHEN {count_sql} > 0 THEN {sql}{default_sql} END)"
+        return sql, count_params + params + default_params
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        if not connection.features.is_postgresql_16:
+            sql, params = super().as_sql(
+                compiler,
+                connection,
+                function="ARRAY_AGG",
+                **extra_context,
+            )
+            return f"TO_JSONB({sql})", params
+        self.template = "%(function)s(%(distinct)s%(expressions)s RETURNING JSONB)"
```

Comment:
Hey @charettes , after applying the above suggestion i get an  error `TypeError:django.db.models.expressions.Func.as_sql() got multiple values for keyword argument 'template'`

Reply:
@lufafajoshua hard to tell for sure because you didn't provide the full traceback but I assume these lines are to blame

https://github.com/django/django/blob/c4f3006aec5d3d9cebcce6d410d1121e5c0a1003/django/db/models/aggregates.py#L128-L137

Looks like there is a bug in there we should do the following instead

```python
extra_context = {
    **extra_context,
    "template": self.filter_template % extra_context.get(
        "template", self.template
    )
}
sql, params = super().as_sql(
    compiler,
    connection,
    filter=filter_sql,
    **extra_context,
)
```

Reply:
This is the full traceback.

`
>File "<console>", line 1, in <module>
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/manager.py", line 87, in manager_method
  return getattr(self.get_queryset(), name)(*args, **kwargs).    
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/query.py", line 604, in aggregate
    return self.query.chain().get_aggregation(self.db, kwargs)        
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/query.py", line 616, in get_aggregation
    result = compiler.execute_sql(SINGLE)            
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 1556, in execute_sql
    sql, params = self.as_sql()                 
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 742, in as_sql
    extra_select, order_by, group_by = self.pre_sql_setup(                                      
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 84, in pre_sql_setup
    self.setup_query(with_col_aliases=with_col_aliases)
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 73, in setup_query
    self.select, self.klass_info, self.annotation_col_map = self.get_select(                                                      
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 296, in get_select
    sql, params = self.compile(col)                
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/sql/compiler.py", line 550, in compile
    sql, params = vendor_impl(self, self.connection)             
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/aggregates.py", line 296, in as_postgresql
    return super().as_sql(compiler, connection, template=template, **extra_context)           
  File "/home/lufafa/Django-Core-Codebase/django/django/db/models/aggregates.py", line 134, in as_sql
    sql, params = super().as_sql(
TypeError: django.db.models.expressions.Func.as_sql() got multiple values for keyword argument 'template'`